### PR TITLE
chore: only warn about KeyError "ast"/"metadata" if debugging

### PR DESCRIPTION
**Pull request by karmacoma-eth**
plus a drive-by simplification of Mapper

### CODE REVIEW

File: src/halmos/__main__.py [hunk 1]
```diff
@@ -1357,30 +1366,19 @@ def parse_build_out(args: HalmosConfig) -> Dict:
...
-                            nodes=[],
+                contract_map[contract_name] = (json_out, contract_type, natspec)
+                parse_symbols(args, contract_map, contract_name)
+
+            except KeyError as err:
+                if str(err) in ["ast", "metadata"]:
+                    # somewhat expected, e.g. for sol files containing only libraries
+                    if args.debug:
+                        debug(
+                            f"Skipped {json_filename} due to parsing failure: "
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:1378]: this is the real change, the rest is a refactoring

File: src/halmos/__main__.py [hunk 2]
```diff
@@ -1357,30 +1366,19 @@ def parse_build_out(args: HalmosConfig) -> Dict:
...
-                    else:
-                        contract_mapping_info.bytecode = bytecode
-
-                    contract_mapping_info = Mapper().get_contract_mapping_info_by_name(
-                        contract_name
-                    )
-                    Mapper().parse_ast(contract_map[contract_name][0]["ast"])
 
-                except Exception:
-                    pass
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:1383]: this is what prompted the refactoring, we don't want to ignore all exceptions

File: src/halmos/__main__.py [hunk 3]
```diff
@@ -1300,8 +1296,24 @@ def mk_arrlen(args: HalmosConfig) -> Dict[str, int]:
     return arrlen
 
 
+def parse_symbols(args: HalmosConfig, contract_map, contract_name):
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:1299]: pulling this in its own method to avoid nested try..except blocks

File: src/halmos/__main__.py [hunk 4]
```diff
@@ -1300,8 +1296,24 @@ def mk_arrlen(args: HalmosConfig) -> Dict[str, int]:
...
+    except Exception:
+        if args.debug:
+            debug(f"error parsing symbols for contract {contract_name}")
+            debug(traceback.format_exc())
+
+
 def parse_build_out(args: HalmosConfig) -> Dict:
-    result = {}  # compiler version -> source filename -> contract name -> (json, type)
+    # compiler version -> source filename -> contract name -> (json, type)
+    result = defaultdict(lambda: defaultdict(dict))
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:1316]: opportunistic, this avoids doing `if x not in y: y[x] = {}` a bunch of times

File: src/halmos/__main__.py [hunk 5]
```diff
@@ -1318,21 +1330,17 @@ def parse_build_out(args: HalmosConfig) -> Dict:
...
+                continue
+            if json_filename.startswith("."):
+                continue
 
-                json_path = os.path.join(sol_path, json_filename)
-                with open(json_path, encoding="utf8") as f:
-                    json_out = json.load(f)
+            json_path = os.path.join(sol_path, json_filename)
+            with open(json_path, encoding="utf8") as f:
+                json_out = json.load(f)
```
Comment (karmacoma-eth) [inline src/halmos/__main__.py:1340]: pulling this out of the try..except block to reduce the scope of the try blocks and minimize nesting

**Comment by karmacoma-eth** [comment]
should address https://github.com/a16z/halmos/issues/281

File: src/halmos/__main__.py (hunk 1 above)
Comment (daejunpark) [reply src/halmos/__main__.py:1378]: what's your intention when err is not ast nor metadata? generate warning, or silently ignore?

**Comment by daejunpark** [comment]
could you separate the main change from the refactoring changes, to make it easier to review the core change? could be a separate pr, or at least a different commit.

File: src/halmos/__main__.py (hunk 1 above)
Comment (daejunpark) [reply src/halmos/__main__.py:1378]: another q: would it be possible to check if the ast / metadata missing error is indeed due to libraries?

Comment (karmacoma-eth) [reply src/halmos/__main__.py:1378]: > another q: would it be possible to check if the ast / metadata missing error is indeed due to libraries?

that was the case every time I looked

Comment (karmacoma-eth) [reply src/halmos/__main__.py:1378]: > what's your intention when err is not ast nor metadata? generate warning, or silently ignore?

oops, the intention is to generate a warning for other errors

Comment (daejunpark) [reply src/halmos/__main__.py:1378]: > > what's your intention when err is not ast nor metadata? generate warning, or silently ignore?
> 
> oops, the intention is to generate a warning for other errors

in that case, i don't think the current code works as intended.

Comment (karmacoma-eth) [reply src/halmos/__main__.py]: fixed in b135251

**Comment by daejunpark** [comment]
[starting a new thread for easier discussion]

> > another q: would it be possible to check if the ast / metadata missing error is indeed due to libraries?
> 
> that was the case every time I looked

ah, my question was, whether it's possible to automatically check if the problematic file is a library or not, and generate the error message only when it is not a library.

indeed, i'm concerned about silently ignoring field-missing errors in non-debug mode. although we haven't seen cases where non-library files are skipped due to missing fields, i believe we still need to allow users review the skipped files, unless we're certain they can be safely skipped.

also, the usability issue described in #281 is not about the warning message itself, but rather that the message is not detailed enough to understand what's happening, and its tone is too strong.

so, my suggestion for the short term is to improve the warning message, allowing users to review them and detect any real issues, without being scared.

a longer-term solution is to implement a mechanism to check whether the given file is a library and ignore the error only when it's a library.

**Comment by karmacoma-eth** [comment]
closing this one, it's getting too messy. Will update the investigation in https://github.com/a16z/halmos/issues/281

**Pull request closed without merging**